### PR TITLE
Experimental skip-ci support for github

### PR DESCRIPTION
**Pull request by tume**
My first lines of Go code so apologies if it is totally out of place :)

I was trying to implement this in the common code as suggested here: https://github.com/bitrise-io/bitrise-webhooks/issues/2 but couldn't realize how to access the commit message from there so decided to put in the TransformRequest instead which seemed like good place because it already has ShouldSkip.

Would like to hear if this is good way to do it or if it should be placed somewhere else? 
Also this only implements it for the github push event so would need to implement it other places also where needed...

### CODE REVIEW

**Comment by viktorbenei** [comment]
Hi,

Thanks for the PR!

A while back I was thinking about where this should be implemented, right now it would be here https://github.com/bitrise-io/bitrise-webhooks/blob/master/service/hook/endpoint.go#L178

At that point you have access to the converted "commit" message, regardless of which service triggered the hook.

The only issue is, how it should respond to the call. E.g. in case the webhook includes more than one commit, one should be skipped but the other one should trigger a build (BitBucket sends batched commits in some cases for example). You can't just return a single "OK, but skipped" like this https://github.com/bitrise-io/bitrise-webhooks/blob/master/service/hook/endpoint.go#L143 .

I think this should still be a "success" (https://github.com/bitrise-io/bitrise-webhooks/blob/master/service/hook/endpoint.go#L181), but with a special message like the "skipped" one.

WDYT?

**Comment by viktorbenei** [comment]
BTW the error in the PR is that you did not handle the `err` returned by the `regex.Compile`:

> service/hook/github/github.go:86:5	r, _ := regexp.Compile(`(?i)\[skip ci\]|\[ci skip\]`)

I'd also suggest you to not to use a regex, a simple [strings.Contains](https://golang.org/pkg/strings/#Contains) should be enough, but of course `regexp` is fine too ;)

**Comment by viktorbenei** [comment]
Additional note for testing: you should install the [Bitrise CLI](https://www.bitrise.io/cli), then you can run `bitrise run test` in the code's root directory (where the `bitrise.yml` file is) to run the full test suite (the PR check runs the same tests).

**Comment by tume** [comment]
Thanks for your comments that sounds like a better approach :). Didn't even realize that you can also receive commits in a batch. I can try redoing my work with that information and see where that leads to.

**Comment by viktorbenei** [comment]
Awesome, let us know if we can help with anything! ;)

**Comment by tume** [comment]
Another try :). 

Had to modify a bit more than wanted in the beginning. Idea was to add new response specific for the skip cases. Not sure if that is what you were thinking but at least this way it should not be problem if the bitbucket sends them in batches.

No tests for this and it now applies the same logic for pull request which might not be desirable.

File: service/hook/endpoint.go [hunk 1]
```diff
@@ -171,10 +172,22 @@ func HTTPHandler(w http.ResponseWriter, r *http.Request) {
 	respondWith := hookCommon.TransformResponseInputModel{
 		Errors:                  []string{},
 		SuccessTriggerResponses: []bitriseapi.TriggerAPIResponseModel{},
+		SkipResponses:           []bitriseapi.SkipAPIResponseModel{},
```
Comment (viktorbenei) [inline service/hook/endpoint.go]: to conform to the naming of the other related params rename this to `SkippedTriggerResponses`

File: service/hook/default_response_provider.go [hunk 2]
```diff
@@ -27,13 +27,19 @@ type SuccessRespModel struct {
...
 	SuccessTriggerResponses []bitriseapi.TriggerAPIResponseModel `json:"success_responses"`
 	FailedTriggerResponses  []bitriseapi.TriggerAPIResponseModel `json:"failed_responses,omitempty"`
 }
 
 // TransformResponse ...
 func (hp DefaultResponseProvider) TransformResponse(input hookCommon.TransformResponseInputModel) hookCommon.TransformResponseModel {
 	httpStatusCode := 201
+
+	if len(input.SuccessTriggerResponses) == 0 && len(input.SkipResponses) > 0 {
+		httpStatusCode = 200
```
Comment (viktorbenei) [inline service/hook/default_response_provider.go]: good idea - please update the README's related section about the possible HTTP status codes the server can return

File: bitriseapi/bitriseapi.go [hunk 3]
```diff
@@ -39,6 +39,13 @@ type TriggerAPIResponseModel struct {
 	TriggeredWorkflow string `json:"triggered_workflow"`
 }
 
+// SkipAPIResponseModel ...
+type SkipAPIResponseModel struct {
+	CommitHash    string `json:"commit_hash"`
```
Comment (viktorbenei) [inline bitriseapi/bitriseapi.go]: I'm not sure whether these params are important at all - the webhook's "request" info should include these, but might be a good idea.

What I'd definitely add is:

```
Message           string `json:"message"`
```

Which would include the description of why it was skipped

File: service/hook/default_response_provider.go (hunk 2 above)
Comment (viktorbenei) [inline service/hook/default_response_provider.go]: actually.. shouldn't we return with a `400` code instead? to make it stand out in the lists, e.g. when you check the webhook history on GitHub and you try to see why a given commit did not start a build?

On the other hand a skip probably shouldn't be `400`, unless there actually was an error somewhere.

WDYT?

File: service/hook/default_response_provider.go [hunk 4]
```diff
@@ -43,6 +49,7 @@ func (hp DefaultResponseProvider) TransformResponse(input hookCommon.TransformRe
 			Errors:                  input.Errors,
 			SuccessTriggerResponses: input.SuccessTriggerResponses,
 			FailedTriggerResponses:  input.FailedTriggerResponses,
+			SkipResponses:  input.SkipResponses,
```
Comment (viktorbenei) [inline service/hook/default_response_provider.go]: please run `go fmt service/...` ;)

File: service/hook/common/common.go [hunk 5]
```diff
@@ -43,6 +43,8 @@ type TransformResponseModel struct {
 type TransformResponseInputModel struct {
 	// Errors include the errors if the build could not trigger
 	Errors []string
+	SkipResponses        []bitriseapi.SkipAPIResponseModel
```
Comment (viktorbenei) [inline service/hook/common/common.go]: to conform to the naming of the other related params rename this to `SkippedTriggerResponses`

**Comment by viktorbenei** [comment]
LGTM, added a couple of things I think should be changed, but I agree with the solution / where and how you implemented the skip functionality ;)

File: service/hook/default_response_provider.go (hunk 2 above)
Comment (tume) [inline service/hook/default_response_provider.go]: This was a bit tricky for me also. I would still consider it to be success when skipped instead of a failure. At least 400 sounds weird for me because it usually means that there were some problems with the request. 

Might be that it should be something else than 200 but didn't really find any better alternative when quickly looking at the available status codes.

Comment (viktorbenei) [inline service/hook/default_response_provider.go]: I agree, most likely `200` is a good code for this case; when the build is actually started the code is `201` anyway, that should be enough indication I think

**Comment by tume** [comment]
- Renamed SkipResponses -> SkippedTriggerResponses. 
- Fixed styling issues.
- Added message to SkipAPIResponseModel just not sure what text to return :) and left commitMessage/has there but can remove if not needed.

File: service/hook/endpoint.go [hunk 6]
```diff
@@ -171,10 +172,23 @@ func HTTPHandler(w http.ResponseWriter, r *http.Request) {
...
+		SkippedTriggerResponses: []bitriseapi.SkipAPIResponseModel{},
 		FailedTriggerResponses:  []bitriseapi.TriggerAPIResponseModel{},
 	}
 	metrics.Trace("Hook: Trigger Builds", func() {
 		for _, aBuildTriggerParam := range hookTransformResult.TriggerAPIParams {
+			commitMessage := aBuildTriggerParam.BuildParams.CommitMessage
+
+			if strings.Contains(commitMessage, "[skip ci]") || strings.Contains(commitMessage, "[ci skip]") {
+				respondWith.SkippedTriggerResponses = append(respondWith.SkippedTriggerResponses, bitriseapi.SkipAPIResponseModel{
+					Message:       "Build skipped by request",
```
Comment (viktorbenei) [inline service/hook/endpoint.go]: Build skipped because the commit message included a skip ci keyword ([skip ci] or [ci skip]).

Comment (tume) [inline service/hook/endpoint.go]: Changed

**Comment by viktorbenei** [comment]
LGTM, just added a comment for the skip response message.

One more thing what would be awesome is to add some tests, as well as adding this to the README. Feel free to add yourself as a contributor too ;)

**Comment by tume** [comment]
If it is looking good otherwise I can try writing couple of test cases during the weekend.

**Comment by viktorbenei** [comment]
It looks perfect, we'll merge it early next week. If you have the time to write tests it would be awesome, otherwise we can merge it as-it-is and write the tests after the merge.

File: service/hook/default_response_provider.go [hunk 7]
```diff
@@ -27,13 +27,19 @@ type SuccessRespModel struct {
 // TransformResponseModel ...
 type TransformResponseModel struct {
 	Errors                  []string                             `json:"errors,omitempty"`
+	SkippedTriggerResponses []bitriseapi.SkipAPIResponseModel    `json:"skipped,omitempty"`
```
Comment (viktorbenei) [inline service/hook/default_response_provider.go]: `json:"skipped,omitempty"` should be `json:"skipped_responses,omitempty"` - to conform to the other response names

File: service/hook/common/common.go [hunk 8]
```diff
@@ -42,7 +42,9 @@ type TransformResponseModel struct {
 // TransformResponseInputModel ...
 type TransformResponseInputModel struct {
 	// Errors include the errors if the build could not trigger
-	Errors []string
+	Errors                  []string
+	SkippedTriggerResponses []bitriseapi.SkipAPIResponseModel
```
Comment (viktorbenei) [inline service/hook/common/common.go]: Add comment - like the one `SuccessTriggerResponses` has and move it to be the last one, after `FailedTriggerResponses`

File: service/hook/endpoint.go [hunk 9]
```diff
@@ -171,10 +172,23 @@ func HTTPHandler(w http.ResponseWriter, r *http.Request) {
...
 		Errors:                  []string{},
 		SuccessTriggerResponses: []bitriseapi.TriggerAPIResponseModel{},
+		SkippedTriggerResponses: []bitriseapi.SkipAPIResponseModel{},
 		FailedTriggerResponses:  []bitriseapi.TriggerAPIResponseModel{},
 	}
 	metrics.Trace("Hook: Trigger Builds", func() {
 		for _, aBuildTriggerParam := range hookTransformResult.TriggerAPIParams {
+			commitMessage := aBuildTriggerParam.BuildParams.CommitMessage
+
+			if strings.Contains(commitMessage, "[skip ci]") || strings.Contains(commitMessage, "[ci skip]") {
```
Comment (viktorbenei) [inline service/hook/endpoint.go]: move this into a function - and write tests for it!

write both "yes, should skip" and "no, this should not be skipped" tests.

Example commit messages to test against:

* `this should be [ci skip]ped`
* `this will not be [CI SKIP]ped`
* `this won't be skipped either: [ ci  skip ]`
* `[skip ci] will skip`
* ...

We use unit tests to document the code, so you should try to add tests to demonstrate both what will be skipped and what won't be skipped, including possible edge cases (e.g. case sensitivity).

File: service/hook/default_response_provider.go (hunk 7 above)
Comment (viktorbenei) [inline service/hook/default_response_provider.go]: and move it to be the last one, after `FailedTriggerResponses`

**Comment by tume** [comment]
Sorry have been a bit busy with other work but will try to fix those tomorrow.

**Comment by viktorbenei** [comment]
Sure, no problem at all, let us know if we can help with anything!

**Comment by viktorbenei** [comment]
Let us know if you need any help!

**Comment by viktorbenei** [comment]
For reference, the related UserVoice feature request https://bitrise.uservoice.com/forums/235233-general/suggestions/10519956-allow-for-configurable-commit-message-pattern-s-t

(we don't plan to make the pattern configurable, at least for now)

**Comment by callmephilip** [comment]
really looking forward for this being merged in 👍

**Comment by nobre84** [comment]
Having this will be awesome! 👍

**Comment by tume** [comment]
Got back from vacation today so can continue where I left last time :)

Fixed code style/comments that you mentioned and will look into tests next after I first get some idea how the tests are done currently.

**Comment by viktorbenei** [comment]
Awesome to hear @tume , just wanted to merge & finish it myself, but it'd be awesome if you could do it instead!

One note: we discussed about whether `[ci skip]` / `[skip ci]` should be searched in the whole commit text, or only in the first line, and we believe that the best solution would be to search for it only in the first line of the commit text.

The reason for this is that e.g. when you do a "squash and merge" here on GitHub, the first line of the commit (title) will be the title of the Pull Request, and the content (other lines of the commit text / commit body) will be a list of all the commits in the PR. It's completely possible that you pushed a commit with "ci skip" during the development of the PR, but of course you don't want to skip the build after the PR is merged.

If we'd search for the pattern in the whole commit text then you should make sure when you squash and merge that you remove it from everywhere in the commit message, while if we only search for the pattern in the first line / title of the commit then it's way easier to spot if you have the pattern in it.

What do you think about this @tume ?

**Comment by viktorbenei** [comment]
For reference, if I'd do a "squash and merge" on this PR right now, without changing the texts auto generated by GitHub, the full commit text (title + body) would look like:

```
Experimental skip-ci support for github (#8)

* Generic skip ci support

* Review fixes

* Changed skipped message

* Improve code comments

```

**Comment by bootstraponline** [comment]
> One note: we discussed about whether [ci skip] / [skip ci] should be searched in the whole commit text, or only in the first line, and we believe that the best solution would be to search for it only in the first line of the commit text.

Pls no. I put `[ci skip]` at the end of the commit text. This is how it's done on Travis. 

> Commits that have [ci skip] or [skip ci] anywhere in the commit messages are ignored by Travis CI.
>
> https://docs.travis-ci.com/user/customizing-the-build#Skipping-a-build

**Comment by viktorbenei** [comment]
@bootstraponline why's putting it into the first line an issue? That would solve the problem/use case I mentioned:

> The reason for this is that e.g. when you do a "squash and merge" here on GitHub, the first line of the commit (title) will be the title of the Pull Request, and the content (other lines of the commit text / commit body) will be a list of all the commits in the PR. It's completely possible that you pushed a commit with "ci skip" during the development of the PR, but of course you don't want to skip the build after the PR is merged.

**Comment by bootstraponline** [comment]
> @bootstraponline why's putting it into the first line an issue? That would solve the problem/use case I mentioned:

1) This is not a real issue, I haven't seen anyone petitioning Travis to adopt a similar fix.
2) The first line is reserved for use by humans, cluttering it up with messages to the CI service is not a good practice.

My typical message is like this:

````
Update selenium-webdriver dependency to 2.48

[ci skip]
````

**Comment by viktorbenei** [comment]
Makes sense, I'm just worried that it's too easy to miss a `[ci skip]` in a merge's commit message, auto generated by a tool - we do this a lot actually, when the commit history is clean and there's no point in changing the commit texts.

**Comment by bootstraponline** [comment]
Let's wait to see if people actually experience this as a legitimate issue.
I think the use case I described is the norm. There should be a very good
reason to break that expectation.

**Comment by viktorbenei** [comment]
I agree, new features are always subject for revisions anyway if it turns out a use case was missed.

**Comment by viktorbenei** [comment]
btw @bootstraponline what about first __and__ last line? Would that work?
I'm personally still not sure whether the middle of the commit text should be checked for the pattern, it's too easy to miss IMO

**Comment by bootstraponline** [comment]
> btw @bootstraponline what about first and last line?  Would that work?

No, that's too magical. If CI skip is present then bitrise shouldn't ignore it.

```
Add some numbers

[ci skip]
Fix more stuff
```

**Comment by viktorbenei** [comment]
@bootstraponline re

```
Add some numbers

[ci skip]
Fix more stuff
```

IMO this commit should be built, as you most likely left `[ci skip]` in it by accident and not by purpose

**Comment by bootstraponline** [comment]
> IMO this commit should be built, as you most likely left [ci skip] in it by accident and not by purpose

Uh, no. It's following the Travis convention which has set the standard. This "by accident" use case doesn't exist.

Is there any CI provider which intentionally disregards users when they explicitly mark a commit as skip?

**Comment by viktorbenei** [comment]
I definitely agree with you @bootstraponline , in general, but given the time we spend in support.. You would be amazed how many times people have issues like this (when they forget to change something and just click, "next next next" style)

**Comment by bootstraponline** [comment]
I would expect you'd get more support issues by being the only vendor in this space to have a non-standard ci skip behavior. They seem to manage support despite having a much larger customer base.

> CircleCI won’t run the build if we find [ci skip] or [skip ci] anywhere in the commit message of the head commit.
>
> https://circleci.com/docs/skip-a-build/

--

> Commits that have [ci skip] or [skip ci] anywhere in the commit messages are ignored by Travis CI.
>
> https://docs.travis-ci.com/user/customizing-the-build/

--

> [skip ci] to the commit message of the last commit before you push and that push will be ignored. 
>
> https://documentation.codeship.com/continuous-integration/skipping-builds/

**Comment by viktorbenei** [comment]
@bootstraponline I'm completely aware how it is in other CI/CD services, but that doesn't mean that there's no better solution, and that we can't discuss about possible use cases where that solution would not work or could be improved

**Comment by bootstraponline** [comment]
I just think something needs to be implemented and then we collect data based on the reality. There's only so much you can figure out without running code.

**Comment by viktorbenei** [comment]
@bootstraponline that I completely agree with ;)

And, why we discuss these last details here is exactly because we plan to merge and release this _real soon_ ;)

**Comment by callmephilip** [comment]
Really looking forward to this one! 🎈

**Comment by viktorbenei** [comment]
@tume added you as a contributor into the README -> https://github.com/bitrise-io/bitrise-webhooks/pull/19/commits/32ce12875f03f88c453872944131f88d028c462c

if you'd want to change that please let me know!

**Comment by viktorbenei** [comment]
Skip CI is now in production, if you use the hooks.bitrise.io server for your webhooks (the default) you can now use `[skip ci]` and `[ci skip]`!

Huge thanks to @tume for the PR, and for everybody else who contributed to the discussion here! :rocket: :tada:

**Comment by nobre84** [comment]
Great stuff ! 🎉 🎉

**Comment by callmephilip** [comment]
yay! 🎆

**Comment by bootstraponline** [comment]
awesome! :shipit:
